File: backend/app/translation/ports.py
from __future__ import annotations

from dataclasses import dataclass, field
from typing import Any, AsyncIterator, Literal, Protocol

Role = Literal["system", "user", "assistant"]
FinishReason = Literal["stop", "length", "error", "cancelled"]


@dataclass(slots=True)
class LLMMessage:
    role: Role
    content: str


@dataclass(slots=True)
class TranslationRequest:
    request_id: str
    source_lang: str
    target_lang: str
    text: str
    context: list[LLMMessage] = field(default_factory=list)
    tone: str | None = None
    glossary: dict[str, str] | None = None
    max_output_tokens: int = 1536
    temperature: float = 0.0
    metadata: dict[str, str] = field(default_factory=dict)


@dataclass(slots=True)
class StreamStart:
    provider: str
    model: str


@dataclass(slots=True)
class StreamDelta:
    text: str


@dataclass(slots=True)
class StreamFinal:
    text: str
    finish_reason: FinishReason
    input_tokens: int | None = None
    output_tokens: int | None = None
    latency_first_token_ms: int | None = None
    latency_total_ms: int | None = None
    cached: bool = False


@dataclass(slots=True)
class StreamError:
    code: str
    message: str
    retryable: bool


LLMStreamEvent = StreamStart | StreamDelta | StreamFinal | StreamError


@dataclass(slots=True)
class ModelCapabilities:
    streaming: bool
    glossary: bool
    json_mode: bool
    prompt_caching: bool
    max_context_tokens: int | None


class TranslationLLM(Protocol):
    provider: str
    model: str
    capabilities: ModelCapabilities

    async def translate_stream(self, req: TranslationRequest) -> AsyncIterator[LLMStreamEvent]:
        ...


class TranslationCache(Protocol):
    async def get(self, key: str) -> dict[str, Any] | None:
        ...

    async def set(self, key: str, value: dict[str, Any], ttl_seconds: int) -> None:
        ...

File: go-app/internal/translation/ports.go
package translation

type LLMMessage struct {
	Role    string `json:"role"`
	Content string `json:"content"`
}

type TranslationRequest struct {
	RequestID       string            `json:"request_id"`
	SourceLang      string            `json:"source_lang"`
	TargetLang      string            `json:"target_lang"`
	Text            string            `json:"text"`
	Context         []LLMMessage      `json:"context"`
	Tone            *string           `json:"tone,omitempty"`
	Glossary        map[string]string `json:"glossary,omitempty"`
	MaxOutputTokens int               `json:"max_output_tokens"`
	Temperature     float64           `json:"temperature"`
	Metadata        map[string]string `json:"metadata"`
}

type StreamStart struct {
	Provider string `json:"provider"`
	Model    string `json:"model"`
}
type StreamDelta struct {
	Text string `json:"text"`
}
type StreamFinal struct {
	Text                string `json:"text"`
	FinishReason        string `json:"finish_reason"`
	InputTokens         *int   `json:"input_tokens,omitempty"`
	OutputTokens        *int   `json:"output_tokens,omitempty"`
	LatencyFirstTokenMS *int   `json:"latency_first_token_ms,omitempty"`
	LatencyTotalMS      *int   `json:"latency_total_ms,omitempty"`
	Cached              bool   `json:"cached"`
}
type StreamError struct {
	Code      string `json:"code"`
	Message   string `json:"message"`
	Retryable bool   `json:"retryable"`
}

type ModelCapabilities struct {
	Streaming        bool
	Glossary         bool
	JSONMode         bool
	PromptCaching    bool
	MaxContextTokens *int
}

type StreamEvent interface{}

type TranslationLLM interface {
	Provider() string
	Model() string
	Capabilities() ModelCapabilities
	TranslateStream(req TranslationRequest) ([]StreamEvent, error)
}

type TranslationCache interface {
	Get(key string) map[string]any
	Set(key string, value map[string]any, ttlSeconds int)
}
